<a href="https://colab.research.google.com/github/Toni2Morales/PrediccionyAnalisisDeVinos/blob/main/src/notebooks/notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predicción y análisis de vinos

## Clonamos el repositorio

In [1]:
!git clone https://github.com/Toni2Morales/PrediccionyAnalisisDeVinos.git

Cloning into 'PrediccionyAnalisisDeVinos'...
remote: Enumerating objects: 25, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (11/11), done.
remote: Total 25 (delta 0), reused 11 (delta 0), pack-reused 9
Unpacking objects: 100% (25/25), 47.14 MiB | 6.84 MiB/s, done.


## Importamos las librerías

In [2]:
import os
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
os.getcwd()

'/content'

In [4]:
os.chdir("PrediccionyAnalisisDeVinos/src/notebooks")
os.getcwd()

'/content/PrediccionyAnalisisDeVinos/src/notebooks'

## Leemos los datos

In [17]:
data1 = pd.read_csv("../data/winemag-data-130k-v2.csv", index_col = 0)
data2 = pd.read_csv("../data/winemag-data_first150k.csv", index_col = 0)
data3 = pd.read_json("../data/winemag-data-130k-v2.json")

In [18]:
data1

country                                        description  \
0          Italy  Aromas include tropical fruit, broom, brimston...   
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967        US  Citation is given as much as a decade of bottl...   
129968    France  Well-drained gravel soil gives this wine its c...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price  \
0                                 Vulkà Bianco      87    NaN   
1                                     Avidagos      87   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2  \
0       Sicily & Sardinia                 Etna                NaN   
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle  \
0            Kerin O’Keefe          @kerinokeefe   
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety  \
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer 

In [19]:
data2

country                                        description  \
0           US  This tremendous 100% varietal wine hails from ...   
1        Spain  Ripe aromas of fig, blackberry and cassis are ...   
2           US  Mac Watson honors the memory of a wine once ma...   
3           US  This spent 20 months in 30% new French oak, an...   
4       France  This is the top wine from La Bégude, named aft...   
...        ...                                                ...   
150925   Italy  Many people feel Fiano represents southern Ita...   
150926  France  Offers an intriguing nose with ginger, lime an...   
150927   Italy  This classic example comes from a cru vineyard...   
150928  France  A perfect salmon shade, with scents of peaches...   
150929   Italy  More Pinot Grigios should taste like this. A r...   

                                 designation  points  price  \
0                          Martha's Vineyard      96  235.0   
1       Carodorum Selección Especial Reserva      96  110.0   
2              Special Selected Late Harvest      96   90.0   
3                                    Reserve      96   65.0   
4                                 La Brûlade      95   66.0   
...                                      ...     ...    ...   
150925                                   NaN      91   20.0   
150926                        Cuvée Prestige      91   27.0   
150927                         Terre di Dora      91   20.0   
150928                       Grand Brut Rosé      90   52.0   
150929                                   NaN      90   15.0   

                  province           region_1           region_2  \
0               California        Napa Valley               Napa   
1           Northern Spain               Toro                NaN   
2               California     Knights Valley             Sonoma   
3                   Oregon  Willamette Valley  Willamette Valley   
4                 Provence             Bandol                NaN   
...                    ...                ...                ...   
150925      Southern Italy  Fiano di Avellino                NaN   
150926           Champagne          Champagne                NaN   
150927      Southern Italy  Fiano di Avellino                NaN   
150928           Champagne          Champagne                NaN   
150929  Northeastern Italy         Alto Adige                NaN   

                   variety                   winery  
0       Cabernet Sauvignon                    Heitz  
1            Tinta de Toro  Bodega Carmen Rodríguez  
2          Sauvignon Blanc                 Macauley  
3               Pinot Noir                    Ponzi  
4       Provence red blend     Domaine de la Bégude  
...                    ...                      ...  
150925         White Blend    Feudi di San Gregorio  
150926     Champagne Blend                H.Germain  
150927         White Blend                Terredora  
150928     Champagne Blend                   Gosset  
150929        Pinot Grigio            Alois Lageder  

[150930 rows x 10 columns]

In [20]:
data3

points                                              title  \
0           87                  Nicosia 2013 Vulkà Bianco  (Etna)   
1           87      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           87      Rainstorm 2013 Pinot Gris (Willamette Valley)   
3           87  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           87  Sweet Cheeks 2012 Vintner's Reserve Wild Child...   
...        ...                                                ...   
129966      90  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...   
129967      90                  Citation 2004 Pinot Noir (Oregon)   
129968      90  Domaine Gresser 2013 Kritt Gewurztraminer (Als...   
129969      90      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)   
129970      90  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...   

                                              description         taster_name  \
0       Aromas include tropical fruit, broom, brimston...       Kerin O’Keefe   
1       This is ripe and fruity, a wine that is smooth...          Roger Voss   
2       Tart and snappy, the flavors of lime flesh and...        Paul Gregutt   
3       Pineapple rind, lemon pith and orange blossom ...  Alexander Peartree   
4       Much like the regular bottling from 2012, this...        Paul Gregutt   
...                                                   ...                 ...   
129966  Notes of honeysuckle and cantaloupe sweeten th...  Anna Lee C. Iijima   
129967  Citation is given as much as a decade of bottl...        Paul Gregutt   
129968  Well-drained gravel soil gives this wine its c...          Roger Voss   
129969  A dry style of Pinot Gris, this is crisp with ...          Roger Voss   
129970  Big, rich and off-dry, this is powered by inte...          Roger Voss   

       taster_twitter_handle  price                             designation  \
0               @kerinokeefe    NaN                            Vulkà Bianco   
1                 @vossroger   15.0                                Avidagos   
2                @paulgwine    14.0                                    None   
3                       None   13.0                    Reserve Late Harvest   
4                @paulgwine    65.0      Vintner's Reserve Wild Child Block   
...                      ...    ...                                     ...   
129966                  None   28.0  Brauneberger Juffer-Sonnenuhr Spätlese   
129967           @paulgwine    75.0                                    None   
129968            @vossroger   30.0                                   Kritt   
129969            @vossroger   32.0                                    None   
129970            @vossroger   21.0           Lieu-dit Harth Cuvée Caroline   

               variety             region_1           region_2  \
0          White Blend                 Etna               None   
1       Portuguese Red                 None               None   
2           Pinot Gris    Willamette Valley  Willamette Valley   
3             Riesling  Lake Michigan Shore               None   
4           Pinot Noir    Willamette Valley  Willamette Valley   
...                ...                  ...                ...   
129966        Riesling                 None               None   
129967      Pinot Noir               Oregon       Oregon Other   
129968  Gewürztraminer               Alsace               None   
129969      Pinot Gris               Alsace               None   
129970  Gewürztraminer               Alsace               None   

                 province   country                                    winery  
0       Sicily & Sardinia     Italy                                   Nicosia  
1                   Douro  Portugal                       Quinta dos Avidagos  
2                  Oregon        US                                 Rainstorm  
3                Michigan        US                                St. Julian  
4                  Oregon        US                             

## Limpieza de datos

In [21]:
data = pd.concat([data1, data2, data3], ignore_index = True)

In [22]:
data

country                                        description  \
0          Italy  Aromas include tropical fruit, broom, brimston...   
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
410867   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
410868        US  Citation is given as much as a decade of bottl...   
410869    France  Well-drained gravel soil gives this wine its c...   
410870    France  A dry style of Pinot Gris, this is crisp with ...   
410871    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price  \
0                                 Vulkà Bianco      87    NaN   
1                                     Avidagos      87   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
410867  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
410868                                    None      90   75.0   
410869                                   Kritt      90   30.0   
410870                                    None      90   32.0   
410871           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2  \
0       Sicily & Sardinia                 Etna                NaN   
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
410867              Mosel                 None               None   
410868             Oregon               Oregon       Oregon Other   
410869             Alsace               Alsace               None   
410870             Alsace               Alsace               None   
410871             Alsace               Alsace               None   

               taster_name taster_twitter_handle  \
0            Kerin O’Keefe          @kerinokeefe   
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
410867  Anna Lee C. Iijima                  None   
410868        Paul Gregutt           @paulgwine    
410869          Roger Voss            @vossroger   
410870          Roger Voss            @vossroger   
410871          Roger Voss            @vossroger   

                                                    title         variety  \
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
410867  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
410868                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
410869  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
410870      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
410871  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer 

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 410872 entries, 0 to 410871
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   country                410741 non-null  object 
 1   description            410872 non-null  object 
 2   designation            290207 non-null  object 
 3   points                 410872 non-null  int64  
 4   price                  379185 non-null  float64
 5   province               410741 non-null  object 
 6   region_1               343318 non-null  object 
 7   region_2               161975 non-null  object 
 8   taster_name            207454 non-null  object 
 9   taster_twitter_handle  197516 non-null  object 
 10  title                  259942 non-null  object 
 11  variety                410870 non-null  object 
 12  winery                 410872 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 40.8+ MB


## Escogiendo la variable target

In [27]:
data.shape

(410872, 10)

In [28]:
len(data["title"].unique())

118842

In [32]:
data.shape[0]/len(data["title"].unique())

3.4572962420693023

#### Hay demasiada varianza en el tipo de vino como para predecirlo con respecto a su descripción, escogamos otro target. Si escogieramos el nombre dle vino para usarlo como nuestra variable target nuestro modelo solo tendría aproximadamente 4 registros para aprender a diferenciar entre todos los tipos de vinos.

In [29]:
len(data["variety"].unique())

758

In [33]:
data.shape[0]/len(data["variety"].unique())

542.0474934036939

## Missing Values

In [24]:
data.isnull().sum()

country                     131
description                   0
designation              120665
points                        0
price                     31687
province                    131
region_1                  67554
region_2                 248897
taster_name              203418
taster_twitter_handle    213356
title                    150930
variety                       2
winery                        0
dtype: int64

#### Las columnas con más valores nulos son region_2 y taster_twitter_handle que son columnas que no nos dan demasiada información relevante así que las vamos a eliminar. También elimnaremos la columna de region_1

In [25]:
data.drop(columns = ["region_2","taster_twitter_handle", "region_1"], inplace = True)

In [26]:
data.isnull().sum()

country           131
description         0
designation    120665
points              0
price           31687
province          131
taster_name    203418
title          150930
variety             2
winery              0
dtype: int64

In [49]:
data.dropna(inplace = True)

In [51]:
data.shape

(108340, 11)

In [52]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 108340 entries, 3 to 410871
Data columns (total 11 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   country      108340 non-null  object 
 1   description  108340 non-null  object 
 2   designation  108340 non-null  object 
 3   points       108340 non-null  int64  
 4   price        108340 non-null  float64
 5   province     108340 non-null  object 
 6   region_1     108340 non-null  object 
 7   taster_name  108340 non-null  object 
 8   title        108340 non-null  object 
 9   variety      108340 non-null  object 
 10  winery       108340 non-null  object 
dtypes: float64(1), int64(1), object(9)
memory usage: 9.9+ MB


#### Ya no tenemos valores nulos. Vamos a comenzar con el análisis de los datos.

## Análisis

### Varianza

In [54]:
data["country"].unique()

array(['US', 'Spain', 'Italy', 'France', 'Argentina', 'Australia',
       'Canada'], dtype=object)

#### Tenemos estos países en donde se realizaron las catas de vinos

In [56]:
len(data["designation"].unique())

23943

In [57]:
len(data["province"].unique())

63

#### Tenemos 63 provincias distintas

In [58]:
len(data["region_1"].unique())

1004

#### Tenemos 1004 regiones distintas donde se realizadon las catas de vinos

In [59]:
len(data["taster_name"].unique())

17

#### Tenemos 17 probadores de vinos

In [60]:
len(data["title"].unique())

48994

#### Tenemos 48994 vinos distintos

In [61]:
len(data["variety"].unique())

434

In [62]:
len(data["winery"].unique())

8583